wgpu-native/src/device.rs

```diff
@@ -210,11 +210,18 @@ impl DestroyedResources<back::Backend> {
             let resource_id = self.mapped.swap_remove(i).value;
             let buf = &buffer_guard[resource_id];
 
-            let usage = match buf.pending_map_operation {
-                Some(BufferMapOperation::Read(..)) => resource::BufferUsageFlags::MAP_READ,
-                Some(BufferMapOperation::Write(..)) => resource::BufferUsageFlags::MAP_WRITE,
-                _ => unreachable!(),
-            };
+            let mut usage = resource::BufferUsageFlags::empty();
+
+            for pending_operation in buf.pending_map_operations.iter() {
```

Comment:
I don't think we can reasonably resolve multiple mapping requests *at once*. At first, we need to resolve each request individually (i.e. only go the the next one when the previous is fully done). Then as an improvement we may try to coalesce read operations together.

Reply:
Ah, so execute the callback and unmap before moving to the next pending operation?

Reply:
How is the async mapping intended to work? I started to make the changes, but I'm still seeing issues where the mapping callback isn't fired unless a command buffer is submitted to a queue prior to recording the render pass. As far as I can tell, something is misbehaving with the submission_index.

For example, if you call `set_bind_group` on the render pass, the buffer never gets populated in the `ready_to_map` array.

Reply:
I think we can have the first implementation simply allowing only a single mapping on a buffer. Once buffer is requested for mapping, we can disallow any other operations on the buffer until the buffer is unmapped.

see https://github.com/gpuweb/gpuweb/blob/master/design/BufferOperations.md#the-gpubuffer-state-machine for mode details about upstream

> As far as I can tell, something is misbehaving with the submission_index.

very possible! Current mapping implementation doesn't work correctly.

> For example, if you call set_bind_group on the render pass, the buffer never gets populated in the ready_to_map array.

Right. It's not going to be "ready" until the GPU is done using it, including as a a part of a bound resource group.

Reply:
I scaled this PR back after coming to conclusion that it was going to be quite a bit more involved. I think part of the complication is that the `submission_index` is updated prior to the triage of buffers which makes the active submission [comparison always evaluate to true](https://github.com/gfx-rs/wgpu/blob/master/wgpu-native/src/device.rs#L226) and `ready_to_map` is never populated. I think the only reason that mapping works without any command recordings is due to the index remaining at the default of `0` because it was never reported as [used](https://github.com/gfx-rs/wgpu/blob/master/wgpu-native/src/device.rs#L1085).

I think moving all aspects of the the buffer state management from the end of the queue submission to the very beginning will be needed to get this working. [I think this is how dawn implemented it](https://dawn.googlesource.com/dawn/+/refs/heads/master/src/dawn_native/vulkan/QueueVk.cpp#31) too.

With that said, I didn't think I'd be able to pull that off with my very rudimentary understanding of how the state tracking works :)

Reply:
Thank you for investigation!

> I think part of the complication is that the submission_index is updated prior to the triage of buffers which makes the active submission comparison always evaluate to true and ready_to_map is never populated

The `triage*` routines are designed to run at *any* point, so this shoudn't matter, in theory. If the user maps a buffer that was just used in a submission recently, surely it will go straight to `&mut a.mapped`, which eventually will be moved to `ready_to_map` when the GPU is done with the submission. There could be other submissions not using the buffer before it's mapped, in which case this code you linked to will not find an active submission, and the `ready_to_map` will be populated right away.

> I didn't think I'd be able to pull that off with my very rudimentary understanding of how the state tracking works

There is great value in spreading that knowledge, and I'll be more than happy to explain. Here we go... a resource can be either in a writable state, or a combination of read states, at any time. We track the state of a resource type by `track::Tracker`, and we have `TrackerSet` that combines all the supported types. We do it on multiple different timelines:
  - device timeline: tracks the global state of all resources between submissions
  - command recording timeline: tracks resources used by a command buffer
  - pass: tracks resources used in a pass

When the timelines meet each other, we merge the trackers. For example, when a render pass is ended, we figure out, for each used resource, what state it should be in. There is no transitions within a render pass. We then append to the command buffer the necessary transitions of those resources. Another example - on submission, we take all the resources used by a command buffer, and we look at the "init" state and "last" state. We then issue transitions between what the device knows about currently and the "init" state, and then update the device tracker with the "last" state.

Now, going back to mapping. Once a resource is requested for mapping, we need to update the device tracker. And until the mapping is done, we need to forbid any operations (on device timeline! stuff can still be recorded into command buffers but not submitted yet) on that resource.